text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+block. You might say, logically, that the mutation occurs inside the `if` block. But if we use this
+definition, then we could not know about the mutation until after monomorphization. Errors generated
+post-monomorphization are generally frowned upon, as it happens quite late in the compilation
+process. But consider this: what if `x` is not actually mutated within the body of `foo`? Now we
+have a window into what actually happens inside the function, and it is something that is not stated
+in the function signature. Not great. In this specific example, it is not even deterministic!
+Because of this, it is quite literally impossible to know whether `x` is actually mutated inside a
+given function. As a result we have no choice: the error _must_ be generated at the point where the
+reference is taken.
+
+Okay, we solved that problem. We know that the mutable use happens on the final line. But what about
+this?
+
+```rust
+let x = Cell::new(5);
+x.set(6);
+```
+
+Rust, as you may know, has [interior mutability]. This is what we are using here. `x` is not
+declared mutable, and it does not need to be. This is the beauty of interior mutability. But it
+introduces a key question: where is the mutation? The answer is that it is **not** a mutation for
+the purposes of this restriction. This is not because the value is not changed: it is. Rather, it is
+the logical result of the semantics of `mut` restrictions and where errors must occur (as described
+after the previous example). If errors are emitted at the point where the mutable reference is
+created, then there can be no such error here, as no mutable reference is ever created. `Cell::set`
+is a method that takes `&self`, not `&mut self`. Surely we can make interior mutability a special
+case, right? No; the only way to work around this is to make any reference to a type with interior
+mutability considered a mutation. Consequently, you could never have a reference to a type
+containing a `mut`-restricted, interior-mutable field. This is unacceptable, so interior mutability
+cannot be considered a mutation for the purposes of this restriction.
+
+[interior mutability]: https://doc.rust-lang.org/reference/interior-mutability.html
+
+### `struct` expressions are not allowed
+
+Given that the most common use for for `mut`-restricted fields is to ensure an invariant, it is
+important that the invariant be enforced. Consider the previous definition of `Time`. If you could
+write
+
+```rust
+Time {
+    hour: 32,
+    minute: 0,
+    second: 0,
+    nanosecond: 0,
+}
+```
+
+then the invariant would be violated, as there are only 24 hours in a day (numbered 0–23). Given
+that the invariant is not enforced by the type system, it cannot be enforced at all in this case. As
+a result, we have no choice but to disallow `struct` expressions for types with `mut`-restricted
+fields. This applies even when [functional update syntax][fru-syntax] is used, as invariants can
```

Comment:
```suggestion
a result, we have no choice but to disallow `struct` expressions for types with `mut`-restricted
fields, in scopes where any fields are read-only. This applies even when [functional update syntax][fru-syntax] is used, as invariants can
```

You cover this below, but I think it would help to mention it here.